src/py/flwr/server/workflow/default_workflows.py

```diff
@@ -0,0 +1,346 @@
+        raise TypeError(f"Expect a LegacyContext, but get {type(context).__name__}.")
+
+    # Retrieve current_round and start_time from the context
+    cfg = context.state.configs_records[CONFIGS_RECORD_KEY]
+    current_round = cast(int, cfg[KEY_CURRENT_ROUND])
+    start_time = cast(float, cfg[KEY_START_TIME])
+
+    # Centralized evaluation
+    parameters = compat.parametersrecord_to_parameters(
+        record=context.state.parameters_records[PARAMS_RECORD_KEY],
+        keep_input=True,
+    )
+    res_cen = context.strategy.evaluate(current_round, parameters=parameters)
+    if res_cen is not None:
+        loss_cen, metrics_cen = res_cen
+        log(
+            INFO,
+            "fit progress: (%s, %s, %s, %s)",
+            current_round,
+            loss_cen,
+            metrics_cen,
+            timeit.default_timer() - start_time,
+        )
+        context.history.add_loss_centralized(server_round=current_round, loss=loss_cen)
+        context.history.add_metrics_centralized(
+            server_round=current_round, metrics=metrics_cen
+        )
+
+
+def default_fit_workflow(driver: Driver, context: Context) -> None:
+    """Execute the default workflow for a single fit round."""
+    if not isinstance(context, LegacyContext):
+        raise TypeError(f"Expect a LegacyContext, but get {type(context).__name__}.")
+
+    # Get current_round and parameters
+    cfg = context.state.configs_records[CONFIGS_RECORD_KEY]
+    current_round = cast(int, cfg[KEY_CURRENT_ROUND])
+    parametersrecord = context.state.parameters_records[PARAMS_RECORD_KEY]
+    parameters = compat.parametersrecord_to_parameters(
+        parametersrecord, keep_input=True
+    )
+
+    # Get clients and their respective instructions from strategy
+    client_instructions = context.strategy.configure_fit(
+        server_round=current_round,
+        parameters=parameters,
+        client_manager=context.client_manager,
+    )
+
+    if not client_instructions:
+        log(INFO, "fit_round %s: no clients selected, cancel", current_round)
+        return
+    log(
+        DEBUG,
+        "fit_round %s: strategy sampled %s clients (out of %s)",
+        current_round,
+        len(client_instructions),
+        context.client_manager.num_available(),
+    )
+
+    # Build dictionary mapping node_id to ClientProxy
+    node_id_to_proxy = {proxy.node_id: proxy for proxy, _ in client_instructions}
+
+    # Build out messages
+    out_messages = [
+        driver.create_message(
+            content=compat.fitins_to_recordset(fitins, True),
+            message_type=MESSAGE_TYPE_FIT,
+            dst_node_id=proxy.node_id,
+            group_id="",
+            ttl="",
+        )
+        for proxy, fitins in client_instructions
+    ]
+
+    # Send instructions to clients and
+    # collect `fit` results from all clients participating in this round
+    messages = list(driver.send_and_receive(out_messages))
+    del out_messages
```

Comment:
One option here would be to replace this part with `fit_clients` from `server.py`. `fit_clients` operates on `ClientProxy` instances. What we get from `strategy.configure_fit` are `DriverClientProxy` instances that already have a reference to our `Driver` object. This means that `fit_clients` should work transparently on these objects.